ouroboros-network/src/Ouroboros/Network/Protocol/PingPong/Codec.hs

```diff
@@ -0,0 +1,53 @@
+{-# LANGUAGE DataKinds           #-}
+{-# LANGUAGE GADTs               #-}
+{-# LANGUAGE KindSignatures      #-}
+{-# LANGUAGE PolyKinds           #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module Ouroboros.Network.Protocol.PingPong.Codec where
+
+import           Control.Monad.Class.MonadST
+
+import           Data.ByteString.Lazy (ByteString)
+
+import qualified Codec.CBOR.Encoding as CBOR (Encoding, encodeWord)
+import qualified Codec.CBOR.Read     as CBOR
+import qualified Codec.CBOR.Decoding as CBOR (Decoder, decodeWord)
+
+import           Network.TypedProtocol.Core
+import           Network.TypedProtocol.Codec
+import           Network.TypedProtocol.PingPong.Type
+
+import           Ouroboros.Network.Codec
+
+codecPingPong
+  :: forall m.
+     ( Monad m
+     , MonadST m
+     )
+  => Codec PingPong CBOR.DeserialiseFailure m ByteString
+codecPingPong = mkCodecCborLazyBS encodeMsg decodeMsg
+ where
+  encodeMsg :: forall (pr :: PeerRole) st st'.
+               PeerHasAgency pr st
+            -> Message PingPong st st'
+            -> CBOR.Encoding
+  encodeMsg (ClientAgency TokIdle) MsgPing = CBOR.encodeWord 0
+  encodeMsg (ServerAgency TokBusy) MsgPong = CBOR.encodeWord 1
+  encodeMsg (ClientAgency TokIdle) MsgDone = CBOR.encodeWord 2
+
+  decodeMsg :: forall (pr :: PeerRole) s (st :: PingPong).
+               PeerHasAgency pr st
+            -> CBOR.Decoder s (SomeMessage st)
+  decodeMsg stok = do
+    key <- CBOR.decodeWord
+    case (stok, key) of
+      (ClientAgency TokIdle, 0) -> return $ SomeMessage MsgPing
+      (ServerAgency TokBusy, 1) -> return $ SomeMessage MsgPong
+      (ClientAgency TokIdle, 2) -> return $ SomeMessage MsgDone
+
+      -- TODO proper exceptions
```

Comment:
This will be returned by the decoder as a `CBOR.DeserialiseFailure`, albeit just one containing that string (and the byte offset when it failed). Not ideal, but we'll need to extend the CBOR lib to do better here.

Reply:
Do you mean that cborg decoder will catch this error and wrap it in `CBOR.DeserialiseFailure`?

Reply:
Yes, it's `fail` in the CBOR decoder monad, not throwing an exception.

Reply:
that's right!